internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -364,6 +374,90 @@ func (r *NumaflowControllerRolloutReconciler) getResourceOperations() (kubeUtil.
+	numaLogger := logger.FromContext(ctx)
+
+	// Try to get the Numaflow Controller Deployment
+	deplObj := kubernetes.GenericObject{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "numaflow-controller",
+			Namespace: controllerRollout.Namespace,
+		},
+	}
+	deployment, err := kubernetes.GetCR(ctx, r.restConfig, &deplObj, "deployments")
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// This could just mean that the deployment has not yet been created.
+			// Log and return to wait for next reconcile call once the deployment is created triggering the reconciliation.
+			numaLogger.Warnf("Numaflow Controller Deployment not found. It may still have to be created.")
+			return nil
+		}
+
+		return fmt.Errorf("error getting the Numaflow Controller Deployment for NumaflowControllerRollout %s/%s: %v", controllerRollout.Namespace, controllerRollout.Name, err)
+	}
+
+	// Parse the Deployment spec of the existing Numaflow Controller Deployment
+	var deploymentSpec appsv1.DeploymentSpec
+	if len(deployment.Spec.Raw) > 0 {
+		err = json.Unmarshal(deployment.Spec.Raw, &deploymentSpec)
+		if err != nil {
+			return fmt.Errorf("unable to parse spec for existing Numaflow Controller Deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+	}
+
+	// Parse the Deployment status of the existing Numaflow Controller Deployment
+	var deploymentStatus appsv1.DeploymentStatus
+	if len(deployment.Status.Raw) > 0 {
+		err = json.Unmarshal(deployment.Status.Raw, &deploymentStatus)
+		if err != nil {
+			return fmt.Errorf("unable to parse status for existing Numaflow Controller Deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+	}
+
+	// Health Check borrowed from argoproj/gitops-engine/pkg/health/health_deployment.go https://github.com/argoproj/gitops-engine/blob/master/pkg/health/health_deployment.go#L27
+	if deployment.Generation <= deploymentStatus.ObservedGeneration {
+		cond := getDeploymentCondition(deploymentStatus, appsv1.DeploymentProgressing)
+		if cond != nil && cond.Reason == "ProgressDeadlineExceeded" {
+			msg := fmt.Sprintf("Deployment %q exceeded its progress deadline", deployment.Name)
+			controllerRollout.Status.MarkChildResourcesUnhealthy("Degraded", msg, controllerRollout.Generation)
+			return nil
+		} else if deploymentSpec.Replicas != nil && deploymentStatus.UpdatedReplicas < *deploymentSpec.Replicas {
+			msg := fmt.Sprintf("Waiting for Deployment rollout to finish: %d out of %d new replicas have been updated...", deploymentStatus.UpdatedReplicas, *deploymentSpec.Replicas)
+			controllerRollout.Status.MarkChildResourcesUnhealthy("Progressing", msg, controllerRollout.Generation)
+			return nil
+		} else if deploymentStatus.Replicas > deploymentStatus.UpdatedReplicas {
```

Comment:
We may never encounter this issue but if deploymentStatus.Replicas could ever be "nil" - there is a check in line #437, will this comparison trigger an exception?